src/tree/tree_model.cc

```diff
@@ -17,158 +22,601 @@ DMLC_REGISTER_PARAMETER(TreeParam);
 namespace tree {
 DMLC_REGISTER_PARAMETER(TrainParam);
 }
-// internal function to dump regression tree to text
-void DumpRegTree(std::stringstream& fo,  // NOLINT(*)
-                 const RegTree& tree,
-                 const FeatureMap& fmap,
-                 int nid, int depth, int add_comma,
-                 bool with_stats, std::string format) {
-  int float_max_precision = std::numeric_limits<bst_float>::max_digits10;
-  if (format == "json") {
-    if (add_comma) {
-      fo << ",";
-    }
-    if (depth != 0) {
-      fo << std::endl;
-    }
-    for (int i = 0; i < depth + 1; ++i) {
-      fo << "  ";
-    }
-  } else {
-    for (int i = 0; i < depth; ++i) {
-      fo << '\t';
+
+/*!
+ * \brief Base class for dump model implementation, modeling closely after code generator.
+ */
+class TreeGenerator {
+ protected:
+  static int32_t constexpr kFloatMaxPrecision =
+      std::numeric_limits<bst_float>::max_digits10;
+  FeatureMap const& fmap_;
+  std::stringstream ss_;
+  bool const with_stats_;
+
+  template <typename Float>
+  static std::string ToStr(Float value) {
+    static_assert(std::is_floating_point<Float>::value,
+                  "Use std::to_string instead for non-floating point values.");
+    std::stringstream ss;
+    ss << std::setprecision(kFloatMaxPrecision) << value;
+    return ss.str();
+  }
+
+  static std::string Tabs(uint32_t n) {
+    std::string res;
+    for (uint32_t i = 0; i < n; ++i) {
+      res += '\t';
     }
+    return res;
   }
-  if (tree[nid].IsLeaf()) {
-    if (format == "json") {
-      fo << "{ \"nodeid\": " << nid
-         << ", \"leaf\": " << std::setprecision(float_max_precision) << tree[nid].LeafValue();
-      if (with_stats) {
-        fo << ", \"cover\": " << std::setprecision(float_max_precision) << tree.Stat(nid).sum_hess;
-      }
-      fo << " }";
-    } else {
-      fo << nid << ":leaf=" << std::setprecision(float_max_precision) << tree[nid].LeafValue();
-      if (with_stats) {
-        fo << ",cover=" << std::setprecision(float_max_precision) << tree.Stat(nid).sum_hess;
-      }
-      fo << '\n';
+
+  static std::string Match(std::string const& input,
+                           std::map<std::string, std::string> replacements) {
+    std::string result = input;
+    for (auto const& kv : replacements) {
+      auto pos = result.find(kv.first);
```

Comment:
Are we assuming that each keyword will occur only once in the template? As it is,
```cpp
Match("{foo} {foo} {bar}", { { "{foo}", "1" }, { "{bar}", "0" } })
```
will only replace the first occurrence of `{foo}` and produce `"1 {foo} 0"`.

Reply:
Yes.  I will add a comment for this, if I place another loop inside then I might need to start optimizing the procedure.

For a real code generator I think treelite is the better place to go unless you want to merge the functionality into XGBoost, then I will put more thoughts into designing a tree compiler.

Reply:
@trivialfis Makes sense. Yes, you should indicate this limitation in the comment.

> For a real code generator I think treelite is the better place to go unless you want to merge the functionality into XGBoost, then I will put more thoughts into designing a tree compiler.

I agree. And I think it's best to keep Treelite as a separate project, given its complexity and scope (e.g. it supports LightGBM).